### PR TITLE
Sash double clicks

### PR DESCRIPTION
**Pull request by mquandalle**
Start implementing #4660.

The provided code defines a new event that is fired when a “stash” is double-clicked, and two usages of it—the first one on the spitted editor frontiers, and the second one on the bottom panel border. It is easier to review this PR commit per commit.

The capture below demonstrates the feature on the spitted editor:

vscode-resize

I may want in implement the “sidebar optimal sizing” that originally motivated #4660, but I would like to have some guidance on the code structure:
- Is `IViewletView` the right place to implement an optional `getOptimalWidth` method? [edit: I’ve chosen `IViewlet` instead]
- What is the best way to perform the HTML elements width measurement? [edit: I’ve found `DOM.getTotalWidth` and `DOM.getRelativeLeft`]
- How do I get the current viewlet in `vs/workbench/layout.ts`? [edit: the [viewlet service](https://github.com/Microsoft/vscode/blob/84a6e40d4356f747c8c3bc841d87d4edc254f7c4/src/vs/workbench/services/viewlet/common/viewletService.ts#L24)]
- Where and how should I write some tests?

Thank you in advance for considering this contribution :-)

### CODE REVIEW

**Comment by bpasero** [comment]
@mquandalle I would just suggest to make the side bar size to its minimal width. Since we never show horizontal scrollbars, every size is good enough to show the contents within. Minimal width is defined in https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/layout.ts#L20

**Comment by mquandalle** [comment]
That would be pretty easy but if you look at Sublime and Atom they both have a smarter implementation that measures the width of the explorer items and uses the larger one as the sidebar width. Do the test! For example if you have an opened sidebar at its “optimal width” and you open some folder that contains longer file names, double clicking on the sidebar border will re-calculate a larger width. As described in #4660 I’d like to have the same behavior on VSCode.

So the idea was to use a static width (like indeed `DEFAULT_MIN_PART_WIDTH`) on the search and debugger viewlets, and let the file-explorer and the git-commiter viewlets implement a method that returns their optimal width.

**Comment by bpasero** [comment]
@mquandalle I see, there is currently no API to compute the optimal width of the explorer tree to prevent truncating of words, but maybe @joaomoreno could chime in on that topic.

**Comment by joaomoreno** [comment]
Renamed *stash* to **sash** in the title.

@bpasero @mquandalle It can be done, but it is non-trivial.

**Comment by mquandalle** [comment]
I got a first version of sidebar children measurement working in 9da61e8.

**Comment by bpasero** [comment]
@mquandalle cool, I think it would make somewhat more sense that every viewlet can implement a function to provide its optimal width. That way, the double click works with all viewlets. If a viewlet decides to not provide optimal width, it should just fallback to the minimal view width as indicated before.

I think you also need to take into account some other aspects: The working files view to the top also has labels that might get truncated. However, I would not expect that the full path of a working file gets revealed, only its label :+1:

**Comment by mquandalle** [comment]
> I think it would make somewhat more sense that every viewlet can implement a function to provide its optimal width. That way, the double click works with all viewlets. If a viewlet decides to not provide optimal width, it should just fallback to the minimal view width as indicated before.

Yes, that’s how I implemented it.

> I think you also need to take into account some other aspects: The working files view to the top also has labels that might get truncated. However, I would not expect that the full path of a working file gets revealed, only its label

Sure, currently I only take into account the explorer tree width but adding the working view as you say makes total sense.

File: src/vs/base/browser/ui/sash/sash.css [hunk 1]
```diff
@@ -23,4 +23,12 @@
...
 .monaco-sash.disabled {
 	cursor: default;
+}
+
+.vertical-cursor-container * {
+	cursor: ew-resize !important;
+}
+
+.horizontal-cursor-container * {
+	cursor: ns-resize !important;
```
Comment (bpasero) [inline src/vs/base/browser/ui/sash/sash.css]: Why is this needed?

File: src/vs/base/browser/ui/sash/sash.ts [hunk 2]
```diff
@@ -112,17 +117,8 @@ export class Sash extends EventEmitter {
...
-			position: 'absolute',
-			top: 0,
-			left: 0,
-			width: '100%',
-			height: '100%',
-			zIndex: 1000000,
-			cursor: this.orientation === Orientation.VERTICAL ? 'ew-resize' : 'ns-resize'
-		});
-
 		let $window = $(window);
```
Comment (bpasero) [inline src/vs/base/browser/ui/sash/sash.ts]: Can you explain why the changes in sash are needed?

File: src/vs/workbench/browser/parts/editor/sideBySideEditorControl.ts [hunk 3]
```diff
@@ -1212,6 +1214,14 @@ export class SideBySideEditorControl extends EventEmitter implements IVerticalSa
 		this.editorActionsToolbar[position].setActions([], [])();
 	}
 
+	private centerStash(a: Position, b: Position): void {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/sideBySideEditorControl.ts]: Please rename all stash => sash

File: src/vs/workbench/common/viewlet.ts [hunk 4]
```diff
@@ -5,4 +5,9 @@
 
 import {IComposite} from 'vs/workbench/common/composite';
 
-export interface IViewlet extends IComposite { }
+export interface IViewlet extends IComposite {
+	/**
+	 * Returns the minimal width needed to avoid any content horizontal truncation
+	 */
+	getOptimalWidth?(): number;
```
Comment (bpasero) [inline src/vs/workbench/common/viewlet.ts]: Suggest to make this a non optional method and implement it by default in the base viewlet class.

File: src/vs/workbench/parts/files/browser/explorerViewlet.ts [hunk 5]
```diff
@@ -180,6 +180,10 @@ export class ExplorerViewlet extends Viewlet {
 		return this.actionRunner;
 	}
 
+	public getOptimalWidth(): number {
+		return this.getExplorerView().getOptimalWidth();
+	}
+
```
Comment (bpasero) [inline src/vs/workbench/parts/files/browser/explorerViewlet.ts]: You need to take the working files view into account too. Note that explorer view can be different when you open no workspace (no folder).

File: src/vs/workbench/parts/files/browser/media/explorerviewlet.css [hunk 6]
```diff
@@ -17,6 +17,10 @@
 	overflow: hidden;
 }
 
+.explorer-viewlet .explorer-item-label a, .explorer-viewlet .working-files-item-label {
+	margin-right: 12px;
+}
+
```
Comment (bpasero) [inline src/vs/workbench/parts/files/browser/media/explorerviewlet.css]: Why?

**Comment by bpasero** [comment]
I added some comments to the code

File: src/vs/base/browser/ui/sash/sash.ts (hunk 2 above)
Comment (mquandalle) [inline src/vs/base/browser/ui/sash/sash.ts]: I’ve written some explanations in the corresponding commit message, 0636b55e2ed8422f16ed7196db1ebdf33b6ac5ed. I can explain more if that’s not clear.

File: src/vs/workbench/common/viewlet.ts (hunk 4 above)
Comment (mquandalle) [inline src/vs/workbench/common/viewlet.ts]: I tried that and I get the following error:

```
*** Error: /home/maxime/github/vscode/src/vs/workbench/browser/parts/sidebar/sidebarPart.ts(81,10): Type 'IComposite' is not assignable to type 'IViewlet'.
  Property 'getOptimalWidth' is missing in type 'IComposite'.
```

which is related to the following code:

```typescript
	public getActiveViewlet(): IViewlet {
		return this.getActiveComposite();
	}
```

How should composites be transformed into viewlets?

File: src/vs/workbench/parts/files/browser/media/explorerviewlet.css (hunk 6 above)
Comment (mquandalle) [inline src/vs/workbench/parts/files/browser/media/explorerviewlet.css]: It’s an additional margin so that the sidebar optimal width isn’t stuck with the end of the largest tree element. I’ll probably move that margin into the code as the CSS margin trick won’t work for the working files optimal width calculation (since we won’t include the path into this calculation).

File: src/vs/workbench/browser/parts/editor/sideBySideEditorControl.ts (hunk 3 above)
Comment (mquandalle) [inline src/vs/workbench/browser/parts/editor/sideBySideEditorControl.ts]: Done

File: src/vs/workbench/parts/files/browser/media/explorerviewlet.css (hunk 6 above)
Comment (mquandalle) [inline src/vs/workbench/parts/files/browser/media/explorerviewlet.css]: Removed.

File: src/vs/workbench/parts/files/browser/explorerViewlet.ts (hunk 5 above)
Comment (mquandalle) [inline src/vs/workbench/parts/files/browser/explorerViewlet.ts]: Done.

**Comment by mquandalle** [comment]
This PR is feature complete.

**Comment by mquandalle** [comment]
f89c156

**Comment by bpasero** [comment]
Thanks, will review next week!

**Comment by bpasero** [comment]
@mquandalle nice stuff, some feedback:
* suggest to move `getOptimalWidth()` into the Viewlet base class since it seems to be used only for viewlets
* for me the implementation of  `getOptimalWidth()` in working files and explorer view is not working (see my video below on mac)
* ideally the implementation of  `getOptimalWidth()` should be shared and reused, would it be possible to make the implementation generic enough that it works with any tree? it could then be moved into `ViewletView` for others to pick up
* once we have a reusable `getOptimalWidth()` I actually think we should use it in search, git and debug too, why would we not?
* the fact that the sash no longer draws a full div for the cursor style seems to cause issues: if you drag the sidebar into the left until it hides and then back you see that the mouse no longer shows the resize cursor (the same is true when you size the panel in and out)

That is it for now :+1:

**Comment by bpasero** [comment]
untitled

**Comment by mquandalle** [comment]
Oops I thought I answered a few hours ago but apparently my message wasn’t sent. On your point 1,

> suggest to move `getOptimalWidth()` into the Viewlet base class since it seems to be used only for viewlets

I initially tried to do that but encountered the following bug: https://github.com/Microsoft/vscode/pull/4702#discussion_r57881357 (GitHub may fold the comment). Actually I don’t understand the distinction between Composite and Viewlet components.

**Comment by mquandalle** [comment]
The bug in point 2, 

> for me the implementation of `getOptimalWidth()` in working files and explorer view is not working (see my video below on mac)

was introduced by a late git history cleaning, it’s now fixed.

**Comment by mquandalle** [comment]
Also note that the current DOM measurement strategy has two important drawbacks:

1. `getOptimalWidth()` doesn’t work if the sidebar is closed, ie you can’t double click on the closed sash to open the sidebar to its optimal width, which is expected;
2. if the explorer list is too long, VS renderer will break the list into multiple “tiles” which won’t be displayed concurrently, thus `getOptimalWidth()` will only consider visible elements and not the entire list.

For comparison both Atom and Sublime handle (2) correctly and doesn’t have problem (1) has they don’t allow sidebar opening using the mouse.

**Comment by mquandalle** [comment]
> the fact that the sash no longer draws a full div for the cursor style seems to cause issues: if you drag the sidebar into the left until it hides and then back you see that the mouse no longer shows the resize cursor (the same is true when you size the panel in and out)

I wasn't able to reproduce this problem on my instance (Ubuntu).

**Comment by bpasero** [comment]
@mquandalle thanks, can you resolve the merge conflicts in this PR? I will have a look soon and provide feedback.

**Comment by bpasero** [comment]
@mquandalle actually much easier to review your change without merging master back in. just let it be :)

**Comment by bpasero** [comment]
@mquandalle getting there, works very nice now. Some more feedback:

* [ ] You can manage to move `getOptimalWidth()` into `IViewlet` by casting to `IViewlet` in the `sidebarPart.ts#getActiveViewlet()` method: `return <IViewlet>this.getActiveComposite();`
This should fix all the compile errors and you can do the move out of `IComposite` into `IViewlet` and related implementations. The default implementation (returning null) can then go into the abstract `Viewlet` class.

* [ ] Be careful in `explorerViewlet#getOptimalWidth()`. you expect the explorer view to be there, but if you don't have a folder open, explorer view is of kind `EmptyView` and does not have any optimal width.

* [ ] the implementation of `getOptimalWidth()` seems very duplicate between `explorerView` and `workingFilesView`. I suggest to create a new method in `DOM` where you pass in the class name of the container you want to measure.

**Comment by mquandalle** [comment]
Thank you for the feedback, I should be able to amend the PR soon. Do you have any comment on https://github.com/Microsoft/vscode/pull/4702#issuecomment-205530593?

**Comment by bpasero** [comment]
@mquandalle sure: 

1. I wonder if double click on the hidden sidebar should actually be a no-op? otherwise I am also fine to leave it as is, though I wonder if this is a use case in common scenarios.

2. that is true and due to the fact that the tree is fully virtual and the DOM only contains what you see. I think that this is OK as is.

**Comment by mquandalle** [comment]
I rebased my branch to `master` to solves the (trivial) merge conflict. I also added a commit to address your points in https://github.com/Microsoft/vscode/pull/4702#issuecomment-207813241.

Points 1 and 3 are fixed, but I have a doubt about 2. You say that I’m not handling the case when `explorerViewlet#getExplorerView()` returns `EmptyView` but the return type of this function is `ExplorerView` and not `ExplorerView | EmptyView`. Should I change that?

**Comment by bpasero** [comment]
@mquandalle changes look good! you are right, getExplorerView() will return only the explorer view, but it can return null in case you did not open a folder (just try to File > Close Folder and doubleclick on the explorer sash).

**Comment by mquandalle** [comment]
Ok, it's fixed.

**Comment by bpasero** [comment]
Thanks so much, merged to master!

**Comment by Tyriar** [comment]
Thanks @mquandalle, I've been wanting this for a while! :fireworks:
